### PR TITLE
Change how we deactivate all revisions

### PR DESCRIPTION
**Pull request by dkurepa**
Improves the deployment command to clean all revisions that might have spawned when we changed the ACA configuration, removes labels before deactivating revisions
https://github.com/dotnet/arcade-services/issues/4103

### CODE REVIEW

**Comment by dkurepa** [comment]
tested in https://dev.azure.com/dnceng/internal/_build/results?buildId=2568839&view=results

**Review by oleksandr-didyk — DISMISSED** [review]

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs [hunk 1]
```diff
@@ -126,30 +126,38 @@ public async Task<int> RunAsync()
         }
     }
 
-    private async Task CleanupRevisionsAsync(IEnumerable<ContainerAppRevisionTrafficWeight> revisionsTrafficWeight)
+    private async Task RemoveRevisionLabel(string revisionName, string? label)
     {
-        // Cleanup all revision labels
-        foreach (var revisionTrafficWeight in revisionsTrafficWeight)
+        if (!string.IsNullOrEmpty(label))
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:131]: Why not just make the function accept `string label` if it does not do anything with `null`? Seems wrong. The compiler would then force you to put the if where you call this which is more natural

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs [hunk 2]
```diff
@@ -126,30 +126,38 @@ public async Task<int> RunAsync()
...
 
-        // Now deactivate all revisions in the list
-        foreach (var revisionTrafficWeight in revisionsTrafficWeight)
+    private async Task CleanupRevisionsAsync(IEnumerable<ContainerAppRevisionTrafficWeight> revisionsTrafficWeight)
+    {
+        List<ContainerAppRevisionResource> activeRevisions = _containerApp.GetContainerAppRevisions()
+            .ToEnumerable()
+            .Where(revision => revision.Data.IsActive == true)
+            .Where(revision => revision.Data.TrafficWeight != 100)
+            .ToList();
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:149]: Why this `ToList`?

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs [hunk 3]
```diff
@@ -126,30 +126,38 @@ public async Task<int> RunAsync()
...
+    }
 
-        // Now deactivate all revisions in the list
-        foreach (var revisionTrafficWeight in revisionsTrafficWeight)
+    private async Task CleanupRevisionsAsync(IEnumerable<ContainerAppRevisionTrafficWeight> revisionsTrafficWeight)
+    {
+        List<ContainerAppRevisionResource> activeRevisions = _containerApp.GetContainerAppRevisions()
+            .ToEnumerable()
+            .Where(revision => revision.Data.IsActive == true)
+            .Where(revision => revision.Data.TrafficWeight != 100)
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:148]:
```suggestion
            .Where(revision => revision.Data.IsActive)
            .Where(revision => revision.Data.TrafficWeight != 100)
```

Comment (premun) [reply src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:148]: Or

```suggestion
            .Where(revision => revision.Data.IsActive && revision.Data.TrafficWeight != 100)
```

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs [hunk 4]
```diff
@@ -126,30 +126,38 @@ public async Task<int> RunAsync()
...
+            .ToList();
+
+        var revisionsToDeactivate = activeRevisions
+            .Select(revision => (
+                revision.Data.Name,
+                revisionsTrafficWeight.FirstOrDefault(trafficWeight => trafficWeight.RevisionName == revision.Data.Name)?.Label))
+            .ToList();
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:155]: Why this `ToList`?

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs (hunk 2 above)
Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs]: it's not needed, true

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs (hunk 3 above)
Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs]: `IsActive` is a `bool?` so had to do a something else. I kind of like having 2 `Wheres` in two lines, feels nicer to look at, I think the compiler does optimization there

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs (hunk 4 above)
Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs]: removed

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs [hunk 5]
```diff
@@ -126,30 +126,36 @@ public async Task<int> RunAsync()
...
     }
 
+    private async Task RemoveRevisionLabel(string revisionName, string label)
+    {
+        var result = await InvokeAzCLI([
+                "containerapp", "revision", "label", "remove",
+                ],
+            [
+                "--label", label
+            ]);
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:136]: The way this is indented seems cursed to me :D

How about this?

```suggestion
        var result = await InvokeAzCLI(
            ["containerapp", "revision", "label", "remove"],
            ["--label", label]);
```

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs [hunk 6]
```diff
@@ -126,30 +126,36 @@ public async Task<int> RunAsync()
...
+        result.ThrowIfFailed($"Failed to remove label {label} from revision {revisionName}. Stderr: {result.StandardError}");
```
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:137]: Isn't the stderr printed in `ThrowIfFailed`?

**Review by premun — APPROVED** [review]

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs (hunk 5 above)
Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:136]: yeah this is cursed

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs (hunk 6 above)
Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:134]: Ah you're right, it prints it out in the `ToString` override

File: src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs [hunk 7]
```diff
@@ -126,30 +126,33 @@ public async Task<int> RunAsync()
         }
     }
 
+    private async Task RemoveRevisionLabel(string revisionName, string label)
+    {
+        var result = await InvokeAzCLI(
+            ["containerapp", "revision", "label", "remove"],
+            ["--label", label]);
+        result.ThrowIfFailed($"Failed to remove label {label} from revision {revisionName}. Stderr: {result.StandardError}");
```
Comment (dkurepa) [inline src/ProductConstructionService/ProductConstructionService.Cli/Operations/DeploymentOperation.cs:134]:
```suggestion
        result.ThrowIfFailed($"Failed to remove label {label} from revision {revisionName}.");
```

**Review by premun — APPROVED** [review]

**Pull request merged**